### PR TITLE
Various changes. Mainly about type restrictions and inference

### PR DESCRIPTION
**Pull request by Evizero**
- remove old readme plot code which moved to FileStorage repo
- allow parameters of different number types
- add tests for type inference
- improve type stability of methods
- remove 0.5 broadcast code (not needed in 0.6)
- pass all arrays to eachindex
- upper bound LearnBase to allow breaking changes

### CODE REVIEW

File: src/elementpenalty.jl [hunk 1]
```diff
@@ -4,69 +4,66 @@ Penalties that are applied element-wise.
...
-function value{T}(p::ElementPenalty, θ::AA{T}, s::AA{T})
+value(p::ElementPenalty, θ::Number, s::Number)       = s * value(p, θ)
+value(p::ElementPenalty, θ::AA{<:Number})            = sum(x -> value(p, x), θ)
+value(p::ElementPenalty, θ::AA{<:Number}, s::Number) = sum(x -> value(p, x, s), θ)
+function value{T <: Number,S <: Number}(p::ElementPenalty, θ::AA{T}, s::AA{S})
     @assert size(θ) == size(s)
-    result = zero(T)
-    for i in eachindex(θ)
-        @inbounds result += value(p, θ[i], s[i])
+    # TODO: make this work: Core.Inference.return_type(value, (typeof(p), T, S))
```
Comment (joshday) [inline src/elementpenalty.jl:13]: What does this mean?

File: src/elementpenalty.jl [hunk 2]
```diff
@@ -118,12 +115,12 @@ immutable ElasticNetPenalty{T <: Number} <: ConvexElementPenalty
         new{T}(α)
     end
 end
-for f in [:value, :deriv]
-    @eval function ($f){T <: Number}(p::ElasticNetPenalty{T}, θ::T)
-        p.α * ($f)(L1Penalty(), θ) + (1 - p.α) * ($f)(L2Penalty(), θ)
+for f in (:value, :deriv)
+    @eval function ($f){T <: Number}(p::ElasticNetPenalty{T}, θ::Number)
+        p.α * ($f)(L1Penalty(), θ) + (one(T) - p.α) * ($f)(L2Penalty(), θ)
```
Comment (joshday) [inline src/elementpenalty.jl:120]: Nitpick/question I don't know the answer to.  Is `one(T)` necessary?

```julia
julia> f(x::Float64) = 1 - x
f (generic function with 1 method)

julia> @code_llvm f(1.0)

define double @julia_f_68409(double) #0 !dbg !5 {
top:
  %1 = fsub double 1.000000e+00, %0
  ret double %1
}
```

**Review by joshday — APPROVED** [review]
This is awesome.  Thanks for working on this!

File: src/elementpenalty.jl (hunk 1 above)
Comment (Evizero) [reply src/elementpenalty.jl:13]: well, right now it uses `result = zero(value(p, first(θ), first(s)))` which calls `value`, even though we just want the return type. I though `Core.inference.return_type` could avoid that. but it doesn't look like it

ref: https://discourse.julialang.org/t/using-core-inference-return-type/2945

File: src/elementpenalty.jl (hunk 2 above)
Comment (Evizero) [reply src/elementpenalty.jl:120]: probably not for the default `Number` types in base. But `Number` can be extended by user packages (for example `DualNumbers`) for which we don't know if `Int` is a neutral type. This seems like the cleanest option to be as user friendly as possible. That said, probably 99% of people use these functions with `Float64`, so I doubt anyone runs into an issue in practice

Comment (joshday) [reply src/elementpenalty.jl:120]: Cool 👍

**Pull request merged**